File: cli/python/src/memgo_cli/backend/json.py
"""递归 JSON 类型和外部数据校验。"""

from __future__ import annotations

import math
from collections.abc import Mapping, Sequence
from typing import TypeAlias

JsonValue: TypeAlias = (
    "bool | int | float | str | Sequence[JsonValue] | Mapping[str, JsonValue] | None"
)
JsonObject: TypeAlias = dict[str, JsonValue]


def json_value(value: object) -> JsonValue:
    """校验并复制 JSON 数据，拒绝不支持的值与非有限数字。"""
    if value is None or isinstance(value, (str, bool, int)):
        return value
    if isinstance(value, float) and math.isfinite(value):
        return value
    if isinstance(value, list):
        return [json_value(item) for item in value]
    if isinstance(value, dict) and all(isinstance(key, str) for key in value):
        return {key: json_value(item) for key, item in value.items()}
    raise ValueError("Expected valid JSON data")


def json_object(value: object) -> JsonObject:
    """取得 JSON 对象，其他顶层类型明确失败。"""
    parsed = json_value(value)
    if not isinstance(parsed, dict):
        raise ValueError("Expected a JSON object")
    return parsed


def json_records(value: object) -> list[JsonObject]:
    """从数组或 results/memories 信封中读取对象列表。"""
    parsed = json_value(value)
    if isinstance(parsed, dict):
        parsed = parsed.get("results", parsed.get("memories"))
    if not isinstance(parsed, list):
        raise ValueError("Expected an array or results/memories array envelope")
    return [json_object(item) for item in parsed]


def json_result(value: object) -> JsonObject | list[JsonObject]:
    """校验添加响应，保留单对象或数组的协议形状。"""
    if isinstance(value, list):
        return [json_object(item) for item in value]
    return json_object(value)


def json_string(value: object, field: str) -> str:
    """校验需要字符串的协议字段，不进行隐式类型转换。"""
    if not isinstance(value, str):
        raise ValueError(f"Invalid field {field}: expected string")
    return value


def json_messages(value: object) -> list[JsonObject]:
    """校验非空消息数组，每项必须包含非空角色和字符串内容。"""
    if not isinstance(value, list) or not value:
        raise ValueError("Invalid messages: expected non-empty array")
    records = [json_object(item) for item in value]
    for index, record in enumerate(records):
        role = json_string(record.get("role"), f"messages[{index}].role")
        if not role:
            raise ValueError(f"Invalid messages[{index}].role: expected non-empty string")
        json_string(record.get("content"), f"messages[{index}].content")
    return records

File: cli/node/src/backend/json.ts
/** 外部 JSON 的开放字段，仅允许可序列化值。 */
export type JsonValue = string | number | boolean | null | JsonObject | JsonValue[];
export interface JsonObject {
  [key: string]: JsonValue | undefined;
}

/**
 * 校验整个 JSON 树，拒绝非 JSON 值；返回值与输入不共享可变对象。
 * @param value - 待处理的字段值。
 * @returns 递归校验后的 JSON 值。
 */
export function parseJsonValue(value: unknown): JsonValue {
  if (value === null || typeof value === 'string' || typeof value === 'boolean') return value;
  if (typeof value === 'number' && Number.isFinite(value)) return value;
  if (Array.isArray(value)) return value.map(parseJsonValue);
  if (typeof value === 'object' && value !== null) {
    return Object.fromEntries(
      Object.entries(value).map(([key, item]) => [key, parseJsonValue(item)])
    );
  }
  throw new Error('Response contains a value that is not valid JSON');
}
/**
 * 对象型协议边界不接受数组、null 或标量。
 * @param value - 待处理的字段值。
 * @returns 通过对象形状校验的 JSON 数据。
 */
export function jsonObject(value: unknown): JsonObject {
  const parsed = parseJsonValue(value);
  if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed))
    throw new Error('Expected a JSON object');
  return parsed;
}
/**
 * 列表接口允许裸数组及已存在的 results/memories 信封，缺少集合时明确失败。
 * @param value - 待处理的字段值。
 * @returns 从数组或列表信封中提取的对象列表。
 */
export function jsonRecords(value: unknown): JsonObject[] {
  const parsed = parseJsonValue(value);
  const records = Array.isArray(parsed)
    ? parsed
    : parsed && typeof parsed === 'object'
      ? (parsed.results ?? parsed.memories)
      : undefined;
  if (!Array.isArray(records))
    throw new Error('Expected an array or a results/memories array envelope');
  return records.map(jsonObject);
}
/**
 * 添加接口兼容对象和对象数组两种既有成功形状。
 * @param value - 待处理的字段值。
 * @returns 保留单对象或对象数组形状的结果。
 */
export function jsonResult(value: unknown): JsonObject | JsonObject[] {
  return Array.isArray(value) ? value.map(jsonObject) : jsonObject(value);
}
